design/paging.md

```diff
@@ -0,0 +1,119 @@
+
+## Nomenclature
+
+* `documents` can be read as documents in in a document based database, like MongoDB or ArangoDB,
+  or rows in a RDBMS, like MySQL or Postgres.
+
+## Options
+
+### Skip and limit / offset and limit
+
+This is the standard paging solution of providing a count of documents to skip and a count
+of documents to return.
+
+#### Pros
+
+* Can page to any location in the data
+* Easily understood by users
+
+#### Cons
+
+* `O(n^2)` time complexity for most database implementations
+  * This means that in larger data sets deep paging gets progressively, exponentially slower
+  * A unlimited skip and limit implementation in the KBase Workspace caused MongoDB to slow to
+    a literal halt due to UI paging code written by someone who should've known better
+    * Skip / limit was removed after that
+  * Often implemented with a maximum skip allowed to prevent slow paging
+    * Elasticsearch's limit is 10K by default
+    * This means that only the data at the front of the sort can be paged
+
+### `start_after` with limit
+
+Provide a string or numerical value after which paging should start. Get the next start after
+value from the last value of the current page.
+
+E.g. if the current page runs from `Ababdeh` to `abaciscus` the `start_after` value for the next
+page would be `abasciscus`.
+
+#### Pros
+
+* Easily understood by users if implemented as a simple "next page" button
+* Users can jump to a particular value rather than having to search through pages if they
+  are allowed to specify the `start_after` value
+  * Essentially a filter used for paging
+* Prior `start_after` values can be cached by the front end to allow for returning to a
+  previous page
+* At worst linear time complexity
+
+#### Cons
+
+* Can only navigate to the next page after the current page. Navigating to page 10 from page
+  1 would require 9 full page loads.
+* May not work well on non-unique fields
+  * If a page has more than a page size's number of identical values in the field being paged over,
+    it is impossible to advance further.
+
+#### Notes
+
+* This is one of the paging strategies the KBase Workspace supports, based on the object Unique
+  Permanent Address
+
+### Precalculated or inherent ordering
+
+In this case, an ordering is applied to every field that may be paged over prior to loading the
+data to a database, or the field has an inherent ordering (like an autoincrementing ID).
+The pages can then be retrieved by providing an order range.
+
+#### Pros
+
+* Can page to any location in the data
+* Easily understood by users
+* At worst linear time complexity
+
+#### Cons
+
+* Any data filtering or deletion disrupts the ordering and may result in pages less than the
+  expected size or entirely empty pages
+
+#### Notes
+
+* This is one of the paging strategies the KBase workspace supports, using the object ID within
+  a particular workspace.
+
+### Calculate and cache paging
```

Comment:
From a UI perspective (ignoring the server side complexity) this is my preferred solution as it can be implemented similarly to skip&limit which is well-supported in our query lib.

Reply:
Well... maybe. 

* You have to provide the query ID every time you page
* It may take a loooong time to return the first page (or in a hybrid system, the first time the full query index is calculated)
* For background index building, you'll have to check and see if a page has been indexed yet before you can actually go there (or the server will response with some sort of `page not ready` error

Reply:
Hmm good point. I may have to build some prototypes and see where the sticky bits are.

Reply:
There's no guarantee we're going to go this route so I wouldn't spend a ton of time on it